File: multiplication/src/main/java/learn/mssb/mssbmultiplication/challenge/ChallengeAttemptRepository.java
package learn.mssb.mssbmultiplication.challenge;

import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;

import java.util.List;

public interface ChallengeAttemptRepository extends CrudRepository<ChallengeAttempt, Long> {

    /**
     * @param userAlias user alias
     * @return the last 10 attempts for a given user, identified by user alias.
     */
    List<ChallengeAttempt> findTop10ByUserAliasOrderByIdDesc(String userAlias);

    /**
     * @param userAlias user alias
     * @return attempts for a given user, identified by user alias.
     */
    @Query("SELECT a FROM ChallengeAttempt a WHERE a.user.alias = ?1 ORDER BY a.id DESC")
    List<ChallengeAttempt> lastAttempts(String userAlias);
}

File: multiplicationkt/src/main/kotlin/challenge/ChallengeAttemptRepository.kt
package learn.mssbkt.multiplicationkt.challenge

import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.CrudRepository
import org.springframework.data.repository.query.Param

interface ChallengeAttemptRepository : CrudRepository<ChallengeAttempt, Long> {
    fun findTop10ByUserAliasOrderByIdDesc(userAlias: String): List<ChallengeAttempt>

    @Query("SELECT a FROM ChallengeAttempt a WHERE a.user.alias = :userAlias ORDER BY a.id DESC")
    fun lastAttempts(@Param("userAlias") userAlias: String): List<ChallengeAttempt>
}
